File: app/src/main/java/com/qimian233/ztool/hook/modules/documentsui/DocumentsUIBypass.java
package com.qimian233.ztool.hook.modules.documentsui;

import android.view.View;
import android.widget.Button;

import com.qimian233.ztool.hook.base.BaseHookModule;

import io.github.libxposed.api.XposedInterface;
import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

/**
 * Android文件选择器(DocumentsUI) 限制解除模块
 * 功能：允许用户在/Android/data等受限目录进行选择操作
 */
public class DocumentsUIBypass extends BaseHookModule {

    public DocumentsUIBypass() {}

    @Override
    public String getModuleName() {
        return "documents_ui_bypass";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                "com.android.documentsui"
        };
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if ("com.android.documentsui".equals(packageName)) {
            if (DEBUG) log("开始加载 DocumentsUI 解除限制模块...");
            hookDocumentInfo(classLoader);
            hookPickFragment(classLoader);
        }
    }

    /**
     * Hook DocumentInfo 类，强制解除目录树选择限制
     */
    private void hookDocumentInfo(ClassLoader classLoader) {
        final String documentInfoClass = "com.android.documentsui.base.DocumentInfo";

        try {
            Class<?> docInfoClass = classLoader.loadClass(documentInfoClass);

            // Hook isBlockedFromTree 方法
            Method isBlockedFromTreeMethod = docInfoClass.getDeclaredMethod("isBlockedFromTree");
            this.xposed.hook(isBlockedFromTreeMethod).intercept(chain -> {
                chain.proceed();
                // 强制返回 false，允许选择所有目录
                return false;
            });
            log("成功 Hook DocumentInfo.isBlockedFromTree");

            // 可选：尝试 Hook isBlocked 方法（部分机型或旧版本存在）
            try {
                Method isBlockedMethod = docInfoClass.getDeclaredMethod("isBlocked");
                this.xposed.hook(isBlockedMethod).intercept(chain -> {
                    chain.proceed();
                    return false;
                });
                log("成功 Hook DocumentInfo.isBlocked");
            } catch (Throwable t) {
                // 方法可能不存在，忽略，不作为主要错误记录
            }

        } catch (Throwable t) {
            logError("Hook DocumentInfo 失败", t);
        }
    }

    /**
     * Hook PickFragment 类，强制启用选择按钮并隐藏遮罩层
     */
    private void hookPickFragment(ClassLoader classLoader) {
        final String pickFragmentClass = "com.android.documentsui.picker.PickFragment";

        try {
            Class<?> pickFragClass = classLoader.loadClass(pickFragmentClass);

            // Hook updateView 方法，在UI更新后强制修改控件状态
            Method updateViewMethod = pickFragClass.getDeclaredMethod("updateView");
            this.xposed.hook(updateViewMethod).intercept(chain -> {
                Object result = chain.proceed();
                Object fragment = chain.getThisObject();

                // 1. 获取并启用 mPick 按钮
                try {
                    Field mPickField = findField(fragment.getClass(), "mPick");
                    if (mPickField != null) {
                        mPickField.setAccessible(true);
                        Object mPick = mPickField.get(fragment);
                        if (mPick instanceof Button) {
                            ((Button) mPick).setEnabled(true);
                        }
                    }
                } catch (NoSuchFieldError e) {
                    // 忽略字段不存在的情况
                }

                // 2. 获取并隐藏 mPickOverlay 覆盖层
                try {
                    Field mPickOverlayField = findField(fragment.getClass(), "mPickOverlay");
                    if (mPickOverlayField != null) {
                        mPickOverlayField.setAccessible(true);
                        Object mPickOverlay = mPickOverlayField.get(fragment);
                        if (mPickOverlay instanceof View) {
                            ((View) mPickOverlay).setVisibility(View.GONE); // View.GONE = 8
                        }
                    }
                } catch (NoSuchFieldError e) {
                    // 忽略字段不存在的情况
                }

                return result;
            });
            log("成功 Hook PickFragment.updateView");

        } catch (Throwable t) {
            logError("Hook PickFragment 失败", t);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/documentsui/DocumentsUIBypass.kt
package com.qimian233.ztool.hook.modules.documentsui

import android.annotation.SuppressLint
import android.view.View
import android.widget.Button
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam

/**
 * Android file picker (DocumentsUI) restriction bypass module.
 * Function: allows the user to select files in restricted directories
 * such as /Android/data.
 */
@SuppressLint("PrivateApi")
class DocumentsUIBypass : AppHookModule() {
    override fun getModuleName(): String = PreferenceKeys.DOCUMENTS_UI_BYPASS.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.DOCUMENTS_UI.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        logger.debug("Loading DocumentsUI restriction bypass module...")
        hookDocumentInfo(classLoader)
        hookPickFragment(classLoader)
    }

    /**
     * Hook the DocumentInfo class to force-remove the directory tree selection restriction.
     */
    private fun hookDocumentInfo(classLoader: ClassLoader) {
        val documentInfoClass = "com.android.documentsui.base.DocumentInfo"

        try {
            val docInfoClass = classLoader.loadClass(documentInfoClass)

            // Hook the isBlockedFromTree method
            val isBlockedFromTreeMethod = docInfoClass.getDeclaredMethod("isBlockedFromTree")
            hookWithId(
                isBlockedFromTreeMethod,
                "is_blocked_from_tree"
            ) { chain ->
                chain.proceed()
                false
            }
            logger.info("Successfully hooked DocumentInfo.isBlockedFromTree")

            // Optional: try hooking the isBlocked method (exists on some devices or older versions)
            try {
                val isBlockedMethod = docInfoClass.getDeclaredMethod("isBlocked")
                hookWithId(isBlockedMethod, "is_blocked") { chain ->
                    chain.proceed()
                    false
                }
                logger.info("Successfully hooked DocumentInfo.isBlocked")
            } catch (_: Throwable) {
                // The method may not exist; ignore, not logged as a main error
            }
        } catch (t: Throwable) {
            logger.error("Failed to hook DocumentInfo", t)
        }
    }

    /**
     * Hook the PickFragment class to force-enable the pick button and hide the overlay.
     */
    private fun hookPickFragment(classLoader: ClassLoader) {
        val pickFragmentClass = "com.android.documentsui.picker.PickFragment"

        try {
            val pickFragClass = classLoader.loadClass(pickFragmentClass)

            // Hook the updateView method to force-modify control states after UI update
            val updateViewMethod = pickFragClass.getDeclaredMethod("updateView")
            hookWithId(updateViewMethod, "update_view") { chain ->
                val result = chain.proceed()
                val fragment = chain.thisObject

                // 1. Get and enable the mPick button
                try {
                    val mPickField = findField(fragment.javaClass, "mPick") // null-safe
                    val mPick = mPickField.get(fragment)
                    if (mPick is Button) {
                        mPick.isEnabled = true
                    }
                } catch (_: NoSuchFieldError) {
                    // Ignore missing field
                }

                // 2. Get and hide the mPickOverlay overlay
                try {
                    val mPickOverlayField =
                        findField(fragment.javaClass, "mPickOverlay") // null-safe
                    val mPickOverlay = mPickOverlayField.get(fragment)
                    if (mPickOverlay is View) {
                        mPickOverlay.visibility = View.GONE // View.GONE = 8
                    }
                } catch (_: NoSuchFieldError) {
                    // Ignore missing field
                }
                result
            }
            logger.info("Successfully hooked PickFragment.updateView")
        } catch (t: Throwable) {
            logger.error("Failed to hook PickFragment", t)
        }
    }
}
